s3fs/core.py

```diff
@@ -1814,14 +1814,55 @@ def __init__(
         self.append_block = False
 
         if "a" in mode and s3.exists(path):
-            loc = s3.info(path)["size"]
+            # See:
+            #  put: https://boto3.amazonaws.com/v1/documentation/api/latest
+            #  /reference/services/s3.html#S3.Client.put_object
+
+            # head : https://boto3.amazonaws.com/v1/documentation/api/latest
+            # /reference/services/s3.html#S3.Client.head_object
+            head = {
+                k: v
+                for k, v in self._call_s3(
+                    "head_object",
+                    self.kwargs,
+                    Bucket=bucket,
+                    Key=key,
+                    **version_id_kw(version_id),
+                    **self.req_kw,
+                ).items()
+                if k
+                in {
+                    "CacheControl",
+                    "ContentDisposition",
+                    "ContentEncoding",
+                    "ContentLanguage",
+                    "ContentLength",
+                    "ContentType",
+                    "Expires",
+                    "WebsiteRedirectLocation",
+                    "ServerSideEncryption",
+                    "SSECustomerAlgorithm",
+                    "SSEKMSKeyId",
+                    "BucketKeyEnabled",
+                    "StorageClass",
+                    "ObjectLockMode",
+                    "ObjectLockRetainUntilDate",
+                    "ObjectLockLegalHoldStatus",
+                    "Metadata",
+                }
+            }
+
+            loc = head.pop("ContentLength")
             if loc < 5 * 2 ** 20:
                 # existing file too small for multi-upload: download
                 self.write(self.fs.cat(self.path))
             else:
                 self.append_block = True
             self.loc = loc
 
+            # Reflect head
+            self.s3_additional_kwargs.update(head)
```

Comment:
Do you think that the user should be able override the values from HEAD at the time of append?

Reply:
I would say yes, user still have choice to let default behavior reflecting head, or updating headers based on new content pushed

Reply:
OK, then this is the wrong way around, the existing values will always win. You would need
```
head.update(self.s3_additional_kwargs)
self.s3_additional_kwargs = head
```

Would be nice to test this too.

Reply:
Just excluded current additional kwargs, if it's ok for you ? Or your solution works too

Reply:
OK